low-level/files/opt/seagate/sspl/setup/resource_map/server.py

```diff
@@ -32,18 +39,133 @@ class ServerMap(ResourceMap):
+                errno.EINVAL,
+                "Health provider doesn't have support for'{rpath}'.")
+        return info
+
+    @staticmethod
+    def get_node_details(node):
+        """
+        Parse node information and returns left string and instance.
+
+        Example
+            "storage"    -> ("storage", "*")
+            "storage[0]" -> ("storage", "0")
+        """
+        res = re.search(r"(\w+)\[([\d]+)\]|(\w+)", node)
+        inst = res.groups()[1] if res.groups()[1] else "*"
+        node = res.groups()[0] if res.groups()[1] else res.groups()[2]
+        return node, inst
+
+    @staticmethod
+    def get_data_template(uid, is_fru: bool):
+        """Returns health template."""
+        return {
+            "uid": uid,
+            "fru": str(is_fru).lower(),
+            "last_updated": int(time.time()),
+            "health": {
+                "status": "",
+                "description": "",
+                "recommendation": "",
+                "specifics": []
+            }
+        }
+
+    @staticmethod
+    def set_health_data(health_data: dict, status, description=None,
+                        recommendation=None, specification=None):
+        """Sets health attributes for a component."""
+        good_state = (status == "OK")
+        if not description:
+            description = "%s %s in good health" % (
+                health_data.get("uid"),
+                'is' if good_state else 'is not')
+        if not recommendation:
+            recommendation = 'None' if good_state\
+                             else "Fault detected, please contact Seagate support."
+
+        health_data["health"].update({
+            "status": status,
+            "description": description,
+            "recommendation": recommendation,
+            "specifics": specification
+        })
+
+    def get_cpu_list(self, mode):
+        """Returns the CPU list as per specified mode."""
+        cpu_info_path = Path(self.cpu_path + mode)
+        # Read the text from /cpu/online file
+        cpu_info = cpu_info_path.read_text()
+        # Drop the \n character from the end of string
+        cpu_info = cpu_info.rstrip('\n')
+        # Convert the string to list of indexes
+        cpu_list = self.sysfs.convert_cpu_info_list(cpu_info)
+        return cpu_list
+
+    def get_cpu_info(self):
+        """Update and return CPU information in specific format."""
+        cpu_data = []
+        cpu_present = self.get_cpu_list("present")
+        cpu_online = self.get_cpu_list("online")
+        cpu_usage = psutil.cpu_percent(interval=None, percpu=True)
+        cpu_usage_dict = dict(zip(cpu_online, cpu_usage))
+
+        for cpu_id in range(0, len(cpu_present)):
+            uid = f"cpu_{cpu_id}"
+            cpu_dict = self.get_data_template(uid, False)
+            online_status = "Online" if cpu_id in cpu_online else "Offline"
+            health_status = "OK" if online_status == "Online" else "Fault"
```

Comment:
Offline CPU cant be stated with health 'Fault' , lets put NA

Reply:
I have put NA for health. I have kept _fault_ in the recommendation message as in CPU fault sensor, we report offline CPUs as faulty.

Reply:
yeah, need to be inline with sensor too, so it seems fine